Negative wall_time exception

<!-- DO NOT DELETE OR IGNORE THIS TEMPLATE.  Failing to adhere to this template and provide the necessary information may lead to your Issue being closed without consideration. -->

## Summary

We got an exception because of a negative wall clock time.

### Steps to reproduce the issue

The exception happens from time to time but we are not able to reproduce it.

### Error Message


```
.venv/lib/python3.11/site-packages/pyomo/contrib/solver/common/base.py:685: in solve
    results: Results = super().solve(model)
                       ^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/pyomo/contrib/solver/solvers/knitro/base.py:96: in solve
    results.timing_info.wall_time = (tock - tick).total_seconds()
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
.venv/lib/python3.11/site-packages/pyomo/common/config.py:2481: in __setattr__
    ConfigDict.__setitem__(self, name, value)
.venv/lib/python3.11/site-packages/pyomo/common/config.py:2443: in __setitem__
    cfg.set_value(val)
.venv/lib/python3.11/site-packages/pyomo/common/config.py:2001: in set_value
    self._setter(value)
.venv/lib/python3.11/site-packages/pyomo/common/config.py:1998: in _setter
    self._data = self._cast(value)
                 ^^^^^^^^^^^^^^^^^
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <pyomo.common.config.UninitializedConfigValue object at 0x7bd28cf95c40>
value = -0.216743

    def _cast(self, value):
        if value is None:
            return value
        if self._domain is not None:
            try:
                if value is not NOTSET:
                    return self._domain(value)
                else:
                    return self._domain()
            except:
                err = sys.exc_info()[1]
                if hasattr(self._domain, '__name__'):
                    _dom = self._domain.__name__
                else:
                    _dom = type(self._domain)
>               raise ValueError(
                    "invalid value for configuration '%s':\n"
                    "\tFailed casting %s\n\tto %s\n\tError: %s"
                    % (self.name(True), value, _dom, err)
                )
E               ValueError: invalid value for configuration 'timing_info.wall_time':
E               	Failed casting -0.216743
E               	to NonNegativeFloat
E               	Error: Expected non-negative float, but received -0.216743

.venv/lib/python3.11/site-packages/pyomo/common/config.py:1658: ValueError
```

### Information on your system

Pyomo version: 6.9.5
Python version: 3.11
Operating system: Linux
How Pyomo was installed (PyPI, conda, source): pip
Solver (if applicable): Knitro

### Additional information

It seems to me that the issue comes from using `datetime.datetime.now(datetime.timezone.utc)` to get the current time, while this is not guaranteed to be monotonic. I suspect these calls should be replaced by `time.monotonic() ` https://docs.python.org/3/library/time.html#time.monotonic or `time.perf_counter()` https://docs.python.org/3/library/time.html#time.perf_counter

`datetime.now` is used at multiple other locations in Pyomo https://github.com/search?q=repo%3APyomo%2Fpyomo%20%22datetime.now%22&type=code

--- a/pyomo/contrib/solver/common/persistent.py
+++ b/pyomo/contrib/solver/common/persistent.py
@@ -11,6 +11,7 @@
 
 import abc
 import datetime
+import time
 from typing import List
 
 from pyomo.core.base.constraint import ConstraintData, Constraint
@@ -508,6 +509,7 @@ class PersistentSolverMixin:
 
     def solve(self, model, **kwds) -> Results:
         start_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        tick = time.perf_counter()
         self._active_config = config = self.config(value=kwds, preserve_implicit=True)
         StaleFlagManager.mark_all_as_stale()
 
@@ -529,9 +531,9 @@ class PersistentSolverMixin:
         res = self._solve()
         self._last_results_object = res
 
-        end_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        tock = time.perf_counter()
         res.timing_info.start_timestamp = start_timestamp
-        res.timing_info.wall_time = (end_timestamp - start_timestamp).total_seconds()
+        res.timing_info.wall_time = tock - tick
         res.timing_info.timer = timer
         self._active_config = self.config
 

--- a/pyomo/contrib/solver/solvers/gurobi_direct.py
+++ b/pyomo/contrib/solver/solvers/gurobi_direct.py
@@ -10,6 +10,7 @@
 #  ___________________________________________________________________________
 
 import datetime
+import time
 import io
 import math
 import operator
@@ -276,6 +277,7 @@ class GurobiDirect(GurobiSolverMixin, SolverBase):
 
     def solve(self, model, **kwds) -> Results:
         start_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        tick = time.perf_counter()
         config = self.config(value=kwds, preserve_implicit=True)
         if not self.available():
             c = self.__class__
@@ -392,9 +394,9 @@ class GurobiDirect(GurobiSolverMixin, SolverBase):
         res.solver_version = self.version()
         res.solver_log = ostreams[0].getvalue()
 
-        end_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        tock = time.perf_counter()
         res.timing_info.start_timestamp = start_timestamp
-        res.timing_info.wall_time = (end_timestamp - start_timestamp).total_seconds()
+        res.timing_info.wall_time = tock - tick
         res.timing_info.timer = timer
         return res
 

--- a/pyomo/contrib/solver/solvers/gurobi_direct_minlp.py
+++ b/pyomo/contrib/solver/solvers/gurobi_direct_minlp.py
@@ -11,6 +11,7 @@
 
 
 import datetime
+import time
 import io
 from operator import attrgetter, itemgetter
 
@@ -592,6 +593,7 @@ class GurobiDirectMINLP(GurobiDirect):
             model (Block): a Pyomo model or Block to be solved
         """
         start_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        tick = time.perf_counter()
         config = self.config(value=kwds, preserve_implicit=True)
         if not self.available():
             c = self.__class__
@@ -656,8 +658,8 @@ class GurobiDirectMINLP(GurobiDirect):
         res.solver_version = self.version()
         res.solver_log = ostreams[0].getvalue()
 
-        end_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        tock = time.perf_counter()
         res.timing_info.start_timestamp = start_timestamp
-        res.timing_info.wall_time = (end_timestamp - start_timestamp).total_seconds()
+        res.timing_info.wall_time = tock - tick
         res.timing_info.timer = timer
         return res

--- a/pyomo/contrib/solver/solvers/ipopt.py
+++ b/pyomo/contrib/solver/solvers/ipopt.py
@@ -13,6 +13,7 @@ import logging
 import os
 import subprocess
 import datetime
+import time
 import io
 import re
 import sys
@@ -344,6 +345,7 @@ class Ipopt(SolverBase):
         "Solve a model using Ipopt"
         # Begin time tracking
         start_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        tick = time.perf_counter()
         # Update configuration options, based on keywords passed to solve
         config: IpoptConfig = self.config(value=kwds, preserve_implicit=True)
         # Check if solver is available
@@ -570,11 +572,9 @@ class Ipopt(SolverBase):
             results.solver_log = ostreams[0].getvalue()
 
         # Capture/record end-time / wall-time
-        end_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        tock = time.perf_counter()
         results.timing_info.start_timestamp = start_timestamp
-        results.timing_info.wall_time = (
-            end_timestamp - start_timestamp
-        ).total_seconds()
+        results.timing_info.wall_time = tock - tick
         results.timing_info.timer = timer
         return results
 

--- a/pyomo/contrib/solver/solvers/knitro/base.py
+++ b/pyomo/contrib/solver/solvers/knitro/base.py
@@ -11,7 +11,8 @@
 
 from abc import abstractmethod
 from collections.abc import Mapping, Sequence
-from datetime import datetime, timezone
+import datetime
+import time
 from io import StringIO
 from typing import Optional
 
@@ -66,7 +67,8 @@ class KnitroSolverBase(SolutionProvider, PackageChecker, SolverBase):
         self._saved_var_values = {}
 
     def solve(self, model: BlockData, **kwds) -> Results:
-        tick = datetime.now(timezone.utc)
+        start_timestamp = datetime.datetime.now(datetime.timezone.utc)
+        tick = time.perf_counter()
         self._check_available()
 
         config = self._build_config(**kwds)
@@ -89,10 +91,10 @@ class KnitroSolverBase(SolutionProvider, PackageChecker, SolverBase):
 
         results = self._postsolve(config, timer)
 
-        tock = datetime.now(timezone.utc)
+        tock = time.perf_counter()
 
-        results.timing_info.start_timestamp = tick
-        results.timing_info.wall_time = (tock - tick).total_seconds()
+        results.timing_info.start_timestamp = start_timestamp
+        results.timing_info.wall_time = tock - tick
         return results
 
     def _build_config(self, **kwds) -> KnitroConfig:

